Sources/Stagehand/Driver/InteractiveDriver.swift

```diff
@@ -0,0 +1,246 @@
+        guard !status.isComplete else {
+            // We're already complete. Nothing to do here.
+            return
+        }
+
+        // Note: Halting currently always counts as "not finishing" the animation, even when relativeTimestamp = 1. We may want to reconsider this at some point. Right now it follows the intention of the method call (halt where you are, don't finish the animation), rather than the visual state (at 1 the animation is visually finished).
+
+        let (relativeTimestamp, didComplete): (Double, Bool) = switch (behavior, mode) {
+        case (.revert, _):
+            (0, false)
+        case let (.halt, .manual(relativeTimestamp)):
+            (relativeTimestamp, false)
+        case let (.halt, .automatic(context)):
+            (context.currentRelativeTimestamp(), false)
+        case (.complete, _):
+            (1, true)
+        }
+
+        if case let .automatic(context) = mode {
+            context.displayLink.invalidate()
+            context.completion?(didComplete)
+        }
+
+        mode = .manual(relativeTimestamp: relativeTimestamp)
+
+        renderCurrentFrame()
+        status = .completed(success: false)
+        completion?(false)
+        animationInstance = nil
+    }
+
+    func animate(
+        to targetRelativeTimestamp: Double,
+        using curve: AnimationCurve = LinearAnimationCurve(),
+        duration: TimeInterval? = nil,
+        completion: ((_ finished: Bool) -> Void)? = nil
+    ) {
+        guard !status.isComplete else {
+            // The animation has already completed, so there's nothing to animate.
+            return
+        }
+
+        // Invalidate any in-progress automatic animation.
+        if case let .automatic(context) = mode {
+            context.displayLink.invalidate()
+            context.completion?(false)
+        }
+
+        let startRelativeTimestamp = lastRenderedFrame?.relativeTimestamp ?? 0
+        let segmentRelativeDuration = (targetRelativeTimestamp - startRelativeTimestamp)
+        let segmentDuration = duration ?? abs(segmentRelativeDuration * endToEndDuration)
+
+        let context = AutomaticContext(
+            displayLink: .init(target: self, selector: #selector(renderCurrentFrame)),
+            startTime: CACurrentMediaTime(),
+            segmentDuration: segmentDuration,
+            segmentCurve: curve,
+            startRelativeTimestamp: startRelativeTimestamp,
+            endRelativeTimestamp: targetRelativeTimestamp,
+            completion: completion
+        )
+
+        mode = .automatic(context)
+
+        context.displayLink.add(to: .main, forMode: .common)
+    }
+
+    func updateProgress(to relativeTimestamp: Double) {
+        guard !status.isComplete else {
+            // The animation has already completed, so there's nothing to animate.
+            return
+        }
+
+        // Invalidate any in-progress automatic animation.
+        if case let .automatic(context) = mode {
```

Comment:
consider writing this as:
```suggestion
        if let context = mode.context {
```

and have a full `switch` within the `context accessor so we don't miss a case in the future?